taglib/dsdiff/dsdifffile.cpp

```diff
@@ -0,0 +1,761 @@
+  if(data.isNull() || data.isEmpty())
+    return;
+
+  // Couldn't find an existing chunk, so let's create a new one.
+  unsigned int i = childChunks.size() - 1;
+  unsigned long offset = childChunks[i].offset + childChunks[i].size + childChunks[i].padding;
+
+  // First we update the global size
+  d->size += (offset & 1) + ((data.size() + 1) & ~1) + 12;
+  insert(ByteVector::fromLongLong(d->size, d->endianness == BigEndian), 4, 8);
+  // And the child chunk size
+  d->chunks[d->childChunkIndex[childChunkNum]].size += (offset & 1) + ((data.size() + 1) & ~1) + 12;
+  insert(ByteVector::fromLongLong(d->chunks[d->childChunkIndex[childChunkNum]].size, d->endianness == BigEndian),
+         d->chunks[d->childChunkIndex[childChunkNum]].offset - 8, 8);
+
+  // Now add the chunk to the file
+  unsigned long long nextRootChunkIdx = length();
+  if((d->childChunkIndex[childChunkNum] + 1) < static_cast<int>(d->chunks.size()))
+    nextRootChunkIdx = d->chunks[d->childChunkIndex[childChunkNum] + 1].offset - 12;
+
+  writeChunk(name, data, offset, std::max<unsigned long long>(0, nextRootChunkIdx - offset), (offset & 1) ? 1 : 0);
+
+  // For root chunks
+  updateRootChunksStructure(d->childChunkIndex[childChunkNum] + 1);
+
+  Chunk64 chunk;
+  chunk.name = name;
+  chunk.size = data.size();
+  chunk.offset = offset + 12;
+  chunk.padding = (data.size() & 0x01) ? 1 : 0;
+
+  childChunks.push_back(chunk);
+}
+
+static bool isValidChunkID(const ByteVector &name)
+{
+  if(name.size() != 4)
+    return false;
+
+  for(int i = 0; i < 4; i++) {
+    if(name[i] < 32 || name[i] > 127)
+      return false;
+  }
+
+  return true;
+}
+
+void DSDIFF::File::updateRootChunksStructure(unsigned int startingChunk)
+{
+  for(unsigned int i = startingChunk; i < d->chunks.size(); i++)
+    d->chunks[i].offset = d->chunks[i - 1].offset + 12 + d->chunks[i - 1].size + d->chunks[i - 1].padding;
+
+  // Update childchunks structure as well
+  if(d->childChunkIndex[PROPChunk] >= static_cast<int>(startingChunk)) {
+    std::vector<Chunk64> &childChunksToUpdate = d->childChunks[PROPChunk];
+    if(childChunksToUpdate.size() > 0) {
+      childChunksToUpdate[0].offset = d->chunks[d->childChunkIndex[PROPChunk]].offset + 12;
+      for(unsigned int i = 1; i < childChunksToUpdate.size(); i++)
+        childChunksToUpdate[i].offset = childChunksToUpdate[i - 1].offset + 12 + childChunksToUpdate[i - 1].size + childChunksToUpdate[i - 1].padding;
+    }
+  }
+  if(d->childChunkIndex[DIINChunk] >= static_cast<int>(startingChunk)) {
+    std::vector<Chunk64> &childChunksToUpdate = d->childChunks[DIINChunk];
+    if(childChunksToUpdate.size() > 0) {
+      childChunksToUpdate[0].offset = d->chunks[d->childChunkIndex[DIINChunk]].offset + 12;
+      for(unsigned int i = 1; i < childChunksToUpdate.size(); i++)
+        childChunksToUpdate[i].offset = childChunksToUpdate[i - 1].offset + 12 + childChunksToUpdate[i - 1].size + childChunksToUpdate[i - 1].padding;
```

Comment:
Lines should be kept at approximately 80 characters (with some flexibility up to 95-ish)